src/main/scala/is/hail/expr/ir/MatrixValue.scala

```diff
@@ -20,214 +19,264 @@ case class MatrixValue(
+  }
 
-    def entriesRVD(): RVD = {
-      val resultStruct = typ.entriesTableType.rowType
-      val fullRowType = typ.rvRowType
-      val localEntriesIndex = typ.entriesIdx
-      val localEntriesType = typ.entryArrayType
-      val localColType = typ.colType
-      val localEntryType = typ.entryType
-      val localNCols = nCols
-
-      val localColValues = colValues.broadcast.value
-
-      rvd.boundary.mapPartitions(resultStruct, { (ctx, it) =>
-        val rv2b = ctx.rvb
-        val rv2 = RegionValue(ctx.region)
-
-        it.flatMap { rv =>
-          val gsOffset = fullRowType.loadField(rv, localEntriesIndex)
-          (0 until localNCols).iterator
-            .filter { i =>
-              localEntriesType.isElementDefined(rv.region, gsOffset, i)
-            }
-            .map { i =>
-              rv2b.clear()
-              rv2b.start(resultStruct)
-              rv2b.startStruct()
-
-              var j = 0
-              while (j < fullRowType.size) {
-                if (j != localEntriesIndex)
-                  rv2b.addField(fullRowType, rv, j)
-                j += 1
-              }
-
-              rv2b.addInlineRow(localColType, localColValues(i).asInstanceOf[Row])
-              rv2b.addAllFields(localEntryType, rv.region, localEntriesType.elementOffsetInRegion(rv.region, gsOffset, i))
-              rv2b.endStruct()
-              rv2.setOffset(rv2b.end())
-              rv2
+  def entriesRVD(): RVD = {
+    val resultStruct = typ.entriesTableType.rowType
+    val fullRowType = typ.rvRowType
+    val localEntriesIndex = typ.entriesIdx
+    val localEntriesType = typ.entryArrayType
+    val localColType = typ.colType
+    val localEntryType = typ.entryType
+    val localNCols = nCols
+
+    val localColValues = colValues.broadcast.value
+
+    rvd.boundary.mapPartitions(resultStruct, { (ctx, it) =>
+      val rv2b = ctx.rvb
+      val rv2 = RegionValue(ctx.region)
+
+      it.flatMap { rv =>
+        val gsOffset = fullRowType.loadField(rv, localEntriesIndex)
+        (0 until localNCols).iterator
+          .filter { i =>
+            localEntriesType.isElementDefined(rv.region, gsOffset, i)
+          }
+          .map { i =>
+            rv2b.clear()
+            rv2b.start(resultStruct)
+            rv2b.startStruct()
+
+            var j = 0
+            while (j < fullRowType.size) {
+              if (j != localEntriesIndex)
+                rv2b.addField(fullRowType, rv, j)
+              j += 1
             }
+
+            rv2b.addInlineRow(localColType, localColValues(i).asInstanceOf[Row])
+            rv2b.addAllFields(localEntryType, rv.region, localEntriesType.elementOffsetInRegion(rv.region, gsOffset, i))
+            rv2b.endStruct()
+            rv2.setOffset(rv2b.end())
+            rv2
+          }
+      }
+    })
+  }
+
+  def insertEntries[PC](makePartitionContext: () => PC, newColType: TStruct = typ.colType,
```

Comment:
here and below are the non-whitespace changes (moved these methods in from MatrixTable